datafusion/core/src/execution/context.rs

```diff
@@ -1642,10 +1642,37 @@ impl SessionState {
 
         match &statement {
             DFStatement::Statement(s) => {
-                visit_relations(s.as_ref(), |relation| {
-                    relations.get_or_insert_with(relation, |_| relation.clone());
-                    ControlFlow::<(), ()>::Continue(())
-                });
+                struct RelationVisitor<'a>(&'a mut hashbrown::HashSet<ObjectName>);
+
+                impl<'a> Visitor for RelationVisitor<'a> {
+                    type Break = ();
+
+                    fn pre_visit_relation(
+                        &mut self,
+                        relation: &ObjectName,
+                    ) -> ControlFlow<()> {
+                        self.0.get_or_insert_with(relation, |_| relation.clone());
+                        ControlFlow::Continue(())
+                    }
+
+                    fn pre_visit_statement(
+                        &mut self,
+                        statement: &Statement,
+                    ) -> ControlFlow<()> {
+                        match statement {
```

Comment:
This is a bit gross, but I hope to split apart the query planning from the other types of query, that will clean this up.